core/src/main/java/org/elasticsearch/index/translog/TranslogService.java

```diff
@@ -118,37 +124,64 @@ private TimeValue computeNextInterval() {
         return new TimeValue(interval.millis() + (ThreadLocalRandom.current().nextLong(interval.millis())));
     }
 
-    private class TranslogBasedFlush implements Runnable {
+    // public for testing
+    public TranslogBasedFlush createTranslogBasedFlush() {
+        return new TranslogBasedFlush();
+    }
+
+    public class TranslogBasedFlush extends AbstractRunnable {
 
         private volatile long lastFlushTime = System.currentTimeMillis();
 
+        // prevent construction
+        private TranslogBasedFlush() {}
+
         @Override
-        public void run() {
+        public void onFailure(Throwable t) {
+            if (t instanceof EsRejectedExecutionException) {
```

Comment:
this also to catch rejections coming from the flush thread pool. See #asyncFlushAndReschedule . I can wrap things there. Probably clearer...